vulnerabilities/importers/ruby.py

```diff
@@ -1,76 +1,126 @@
-            'fixed_versions': unaffected_versions,
-            'affected_versions': affected_versions,
-            'advisory': advisory_url
-        })
-    return vulnerability_package_dicts
+from dephell_specifier.range_specifier import InvalidSpecifier
+from packageurl import PackageURL
+import requests
+import yaml
+
+from vulnerabilities.data_source import Advisory
+from vulnerabilities.data_source import GitDataSource
+
+
+class rubyDataSource(GitDataSource):
+
+    def __enter__(self):
+        super(rubyDataSource, self).__enter__()
+
+        if not getattr(self, '_added_files', None):
+            self._added_files, self._updated_files = self.file_changes(
+                recursive=True, file_ext='yml', subdir='./gems')
+
+    def updated_advisories(self) -> Set[Advisory]:
+        files = self._updated_files.union(self._added_files)
+        advisories = []
+        for f in files:
+            processed_data = self.process_file(f)
+            if processed_data:
+                advisories.append(processed_data)
+        return self.batch_advisories(advisories)
+
+    def process_file(self, path) -> List[Advisory]:
+        with open(path) as f:
+            record = yaml.safe_load(f)
+            package_name = record.get(
+                'gem')
+
+            if not package_name:
+                return
+
+            if 'cve' in record:
+                cve_id = 'CVE-{}'.format(record['cve'])
+            else:
+                return
+
+            safe_version_ranges = record.get('patched_versions', [])
+            # this case happens when the advisory contain only 'patched_versions' field
+            # and it has value None(i.e it is empty :( ).
+            if not safe_version_ranges:
+                safe_version_ranges = []
+            safe_version_ranges += record.get('unaffected_versions', [])
+            safe_version_ranges = [i for i in safe_version_ranges if i]
+
+            if not getattr(self, 'pkg_manager_api', None):
+                self.pkg_manager_api = rubyAPI()
+            all_vers = self.pkg_manager_api.get_all_version_of_package(
+                package_name)
+            safe_versions, affected_versions = self.categorize_versions(
+                all_vers, safe_version_ranges)
+
+            impacted_purls = {
+                PackageURL(
+                    name=package_name,
+                    type='gem',
+                    version=version,
+                ) for version in affected_versions}
+
+            resolved_purls = {
+                PackageURL(
+                    name=package_name,
+                    type='gem',
+                    version=version,
+                ) for version in safe_versions}
+
+            return Advisory(
+                summary=record.get('description', ''),
+                impacted_package_urls=impacted_purls,
+                resolved_package_urls=resolved_purls,
+                reference_urls=record.get('url', ''),
```

Comment:
`reference_urls` should be a list or other `Sequence` type.

Reply:
That is huge blunder(on my part), my db is filled with single letter 'urls',  was this the cause ?

Reply:
Probably. But even if there was some kind of checking of the type annotations, it might not have been caught, since `str` might be a `Sequence` of characters. So don't be too hard on yourself. :) It's not even merged yet, so easy to fix.

Reply:
Working fine now.